onadata/apps/api/urls/v2_urls.py

```diff
@@ -2,9 +2,70 @@
 """
 Custom rest_framework Router V2
 """
+from rest_framework import routers
+
 from onadata.apps.api.viewsets.v2.tableau_viewset import TableauViewSet
+from onadata.apps.api.viewsets.v2.imports_viewset import ImportsViewSet
 
 from .v1_urls import MultiLookupRouter
 
-router = MultiLookupRouter(trailing_slash=False)
-router.register(r"open-data", TableauViewSet, basename="open-data")
+
+class DetailedPostRouter(routers.DefaultRouter):
+    """
+    Custom router
+    """
+    routes = [
+        # List route.
+        routers.Route(
+            url=r"^{prefix}{trailing_slash}$",
+            mapping={"get": "list", "post": "create"},
+            name="{basename}-list",
+            detail=False,
+            initkwargs={"suffix": "List"},
+        ),
+        # Dynamically generated list routes. Generated using
+        # @action(detail=False) decorator on methods of the viewset.
+        routers.DynamicRoute(
+            url=r"^{prefix}/{url_path}{trailing_slash}$",
+            name="{basename}-{url_name}",
+            detail=False,
+            initkwargs={},
+        ),
+        # Detail route.
+        routers.Route(
+            url=r"^{prefix}/{lookup}{trailing_slash}$",
+            mapping={
+                "get": "retrieve",
+                "put": "update",
+                "patch": "partial_update",
+                "delete": "destroy",
+                "post": "create",
+            },
+            name="{basename}-detail",
+            detail=True,
+            initkwargs={"suffix": "Instance"},
+        ),
+        # Dynamically generated detail routes. Generated using
+        # @action(detail=True) decorator on methods of the viewset.
+        routers.DynamicRoute(
+            url=r"^{prefix}/{lookup}/{url_path}{trailing_slash}$",
+            name="{basename}-{url_name}",
+            detail=True,
+            initkwargs={},
+        ),
+    ]
+
+    # pylint: disable=redefined-outer-name
+    def extend(self, router):
+        """
+        Extends the routers routes with the routes from another router
+        """
+        self.registry.extend(router.registry)
+
+
+base_router = MultiLookupRouter(trailing_slash=False)
+router = DetailedPostRouter(trailing_slash=False)
+base_router.register(r"open-data", TableauViewSet, basename="open-data")
+router.register(r"imports", ImportsViewSet, basename="imports")
+
+router.extend(base_router)
```

Comment:
Why is the DetailedPostRouter necessary? Which path is impossible to do with the existing default and custom routers that we need another custom router?

Reply:
This was mostly implemented so that we could pass a detail to a POST route; AFAIK the post (create) route on DRF doesn't usually support detail... Is there a particular way we could implement this without having to create our own custom route?

tl;dr This was done so that we could pass detail to the default `create` function; Allowing requests such as `curl -X POST "http://<domain>/api/v2/imports/<xform_id>"`

Reply:
We should use create/post on the `/api/v2/imports` to create a new import instead of `/api/v2/imports/<xform_id>`. The `xform_id` should be a mandatory post parameter in the post to create an import. With that approach, there is no need for a custom router. We already take this approach with the `/api/v1/forms`, `/api/v1/dataviews` and other endpoints.